### PR TITLE
Some miscellaneous fixes for next release (Word8 module)

### LINKED ISSUE #2: intersperse doesn't work
`intersperse` ends up just adding the bytestring to every element, rather than everything but the last element. There is a comment in the code that it "isn't quite right", which appears to be true.

### PR DESCRIPTION
**Pull request by vdukhovni**
- Add missing signature comments to export list, either they should all be
  present (or else perhaps none?)
- More performant rewrite of `denull`.  Less abstract machinery.
- Fix intersperse implementation to handle empty chunks.
- Fix intercalate implementation to not insert anything between
  the final substream and the outer stream end.
- Import 'SPEC' from GHC.Types, as of 9.0, GHC.Exts no longer
  exports `SpecConstrAnnotation`.

### CODE REVIEW

File: lib/Streaming/ByteString.hs [hunk 1]
```diff
@@ -56,18 +57,17 @@ module Streaming.ByteString
   , pack             -- pack :: Monad m => Stream (Of Word8) m r -> ByteStream m r
   , unpack           -- unpack :: Monad m => ByteStream m r -> Stream (Of Word8) m r
   , fromLazy         -- fromLazy :: Monad m => ByteString -> ByteStream m ()
-  , toLazy           -- toLazy :: Monad m => ByteStream m () -> m ByteString
-  , toLazy_          -- toLazy' :: Monad m => ByteStream m () -> m (Of ByteString r)
+  , toLazy           -- toLazy :: Monad m => ByteStream m () -> m (Of ByteString r)
+  , toLazy_          -- toLazy_ :: Monad m => ByteStream m () -> m ByteString
```
Comment (fosskers) [inline lib/Streaming/ByteString.hs]: You know what, let's just kill all these signature comments. They're a maintenance burden.

File: lib/Streaming/ByteString.hs [hunk 2]
```diff
@@ -410,15 +423,44 @@ testNull p@(Chunk bs rest) = if B.null bs
...
+    -- non-empty chunk is found, just apply the loop to the next substream in
+    -- the terminal value via fmap.  If Empty comes up before that happens,
+    -- continue the current stream instead with its denulled tail.
+    --
+    -- This implementation is tail recursive:
+    -- * Recursion via 'loop . Left' continues scanning an inner ByteStream.
+    -- * Recursion via 'loop . Right' moves to the next substream.
+    --
+    -- The old version below was shorter, but noticeably slower, especially
+    -- when empty substreams are frequent:
```
Comment (fosskers) [inline lib/Streaming/ByteString.hs:437]: How did you measure the perf improvement?

File: lib/Streaming/ByteString/Char8.hs [hunk 3]
```diff
@@ -514,17 +516,13 @@ lines text0 = loop1 text0
...
   loop str =  case str of
     Return r -> Empty r
     Step bstr   -> do
       st <- bstr
-      let bs = unlines st
-      case bs of
-        Chunk "" (Empty r)   -> Empty r
-        Chunk "\n" (Empty _) -> bs
-        _                    -> cons' '\n' bs
+      cons' '\n' $ unlines st
```
Comment (fosskers) [inline lib/Streaming/ByteString/Char8.hs:531]: What was the issue here?

File: lib/Streaming/ByteString/Internal.hs [hunk 4]
```diff
@@ -316,22 +312,39 @@ chunkOverhead = 2 * sizeOf (undefined :: Int)
 --             assert (l' <= l) $ return (B.PS fp 0 l', res)
 -- {-# INLINABLE packBytes' #-}
 
+-- | Convert a `Stream` of pure `Word8` into a chunked 'ByteStream'.
 packBytes :: Monad m => Stream (Of Word8) m r -> ByteStream m r
 packBytes cs0 = do
+  -- XXX: Why 32?  It seems like a rather small chunk size, wouldn't
+  -- smallChunkSize make a better choice?
```
Comment (fosskers) [inline lib/Streaming/ByteString/Internal.hs:320]: Weren't the chunks supposed to be 32 *kilobytes*?

File: lib/Streaming/ByteString/Internal.hs [hunk 5]
```diff
@@ -447,7 +460,7 @@ foldrChunksM step nil bs = dematerialize bs (const nil) step join
 unfoldrNE :: Int -> (a -> Either r (Word8, a)) -> a -> (B.ByteString, Either r a)
 unfoldrNE i f x0
     | i < 0     = (B.empty, Right x0)
-    | otherwise = unsafePerformIO $ B.createAndTrim' i $ \p -> go p x0 0
+    | otherwise = B.accursedUnutterablePerformIO $ B.createAndTrim' i $ \p -> go p x0 0
```
Comment (fosskers) [inline lib/Streaming/ByteString/Internal.hs]: Weird, I'm surprised I missed that.

File: lib/Streaming/ByteString.hs (hunk 1 above)
Comment (vdukhovni) [reply lib/Streaming/ByteString.hs]: That's fine.  Would you like me to do that in this PR?

File: lib/Streaming/ByteString.hs (hunk 2 above)
Comment (vdukhovni) [reply lib/Streaming/ByteString.hs:437]: I ran two tests:
- One with a large file with no empty lines, where `lines` produces a stream with no *nulls*. In this case runtime for the old `denull` algorithm was approximately 1.5x the new runtime.
- A second with a large file where 90% of the lines were empty, so that `denull` eliminates most lines. Here the runtime for the old `denull` was around 4-5x the runtime of the new.

The code I was timing was (new):
```haskell
module Main (main) where
import qualified Streaming.ByteString as Q
import qualified Streaming.ByteString.Char8 as Q8

main :: IO ()
main = Q.stdout
    $ Q.intercalate (Q.singleton 0x0a)
    $ Q.denull
    $ Q8.lines
    $ Q.stdin
```
vs. (old):
```haskell
module Main (main) where
import qualified Streaming.ByteString as Q
import qualified Streaming.ByteString.Char8 as Q8
import Streaming.ByteString.Internal (ByteStream(..))
import Streaming.Internal (Stream(..))
import qualified Streaming.Prelude as S
import qualified Streaming as S

denull :: Monad m => Stream (ByteStream m) m r -> Stream (ByteStream m) m r
denull = S.hoist (S.run . S.maps Q.effects) . S.separate . S.mapped Q.nulls

main :: IO ()
main = Q.stdout
    $ Q.intercalate (Q.singleton 0x0a)
    $ denull
    $ Q8.lines
    $ Q.stdin
```

File: lib/Streaming/ByteString/Char8.hs (hunk 3 above)
Comment (vdukhovni) [reply lib/Streaming/ByteString/Char8.hs:531]: It is explained in the commit message. Simply put, `unlines` needs to put back a newline after each substeam, unconditionally. It does not matter what "form" of empty the stream takes. They're all semantically equivalent, and the distinctions made here are fragile and dubious, why is `Chunk "" (Empty r)` somehow different from `Empty r` or from `Go $ pure (Empty r)` or `Go $ pure $ Chunk "" (Empty r)` or ...

It is all just wrong.  It is actually non-trivial to even construct examples where the original pattern matches happen, but it is possible, and they do the wrong thing.

File: lib/Streaming/ByteString/Internal.hs (hunk 4 above)
Comment (vdukhovni) [reply lib/Streaming/ByteString/Internal.hs:320]: I don't know.  What I do know is that they were 32 bytes, for no obvious reason.

**Comment by vdukhovni** [comment]
This library is small, so I pretty much just speed read the whole thing, and tackled anything that caught my eye as looking a bit odd.  Many of the functions are plain enough to not require any attention, but a few stood out...

File: lib/Streaming/ByteString/Char8.hs (hunk 3 above)
Comment (vdukhovni) [reply lib/Streaming/ByteString/Char8.hs:531]: [ TL;DR, see the "TLDR" heading near the bottom, you can skip some of the comments in the middle on first reading if you like ]
I've found the commit that [introduced](https://github.com/haskell-streaming/streaming-bytestring/commit/a8fe89f5aeb3fd30ba85dfe5eb706d3a6caac8a2#diff-9f1312b235f4ef67d4d8608574595e4aae11d7cd75f98bdd784527c506e0d320L416-R429) the incorrect behaviour, the special cases just look like a mistake, there is no stated rationale for introducing them, and the original code was closer to being correct.

I'm essentially returning to the original implementation, but it too had a bug, it was incorrectly turning an empty stream (of bytestream lines) into a singleton stream containing one line, but that's wrong.  Empty streams must stay empty, an empty stream is not the same as a stream with a singline empty line.

Perhaps that original bug (and trying to deal with it in some way) is the reason for the complex cases in the new implementation, but both are wrong.

Basically, given a stream of lines (as `Stream (ByteStream m) m r`) in which the newline ending had been removed from each one, our job is just to put a newline after each line, which can be done by prepending a newline to the result of `unlines` on the next stream, and doing nothing when we reach the end of the outer stream, i.e. `Return r -> Empty r`.

This way:
- `Return r -> Empty r`.  No lines in => no lines out.
- The last line gets its newline ending prepended to the final `Empty r`.
- Same thing for all previous lines, they prepend a newline to the stream obtained by calling `unlines` on the remaining lines.

Because lines 525 and 526 were pattern matching the output of `unlines`, it is useful t consider the conditions under which this same `unlines` can produce either pattern in from some `remaining lines in the stream`. For that we need to peek at the [definition](https://github.com/haskell-streaming/streaming-bytestring/commit/a8fe89f5aeb3fd30ba85dfe5eb706d3a6caac8a2#diff-9f1312b235f4ef67d4d8608574595e4aae11d7cd75f98bdd784527c506e0d320L416-R429)  of `cons'`:
```haskell
cons' :: Char -> ByteStream m r -> ByteStream m r
cons' c (Chunk bs bss) | B.length bs < 16 = Chunk (B.cons (c2w c) bs) bss
cons' c cs             = Chunk (B.singleton (c2w c)) cs
```
which merges the prepended newline with the next chunk (when the stream starts immediately with a Chunk) if it is shorted than 16 bytes, and otherwise prepends a newline as a singleton.

The `unlines` prior to this PR reads:
```haskell
unlines :: Monad m => Stream (ByteStream m) m r ->  ByteStream m r
unlines = loop where
  loop str =  case str of
    Return r -> Empty r
    Step bstr   -> do
      st <- bstr
      let bs = unlines st
      case bs of
        Chunk "" (Empty r)   -> Empty r
        Chunk "\n" (Empty _) -> bs
        _                    -> cons' '\n' bs
    Effect m  -> Go (fmap unlines m)
```
- The base case is an empty stream`Return r`, it is correctly sent to the Empty bytestream `Empty r`.
- We could also an empty stream with monadic effects, such as `Effect $ return $ Return r`. This is mapped to `Go $ return $ Empty r`, which has equivalent content to the simpler base case, but notice how the special-case pattern matches would no longer be able to match this, introducing artificial distinctions between what should be equivalent streams. The placement of effects in the stream should not alter the behaviour of `unlines`.
- The next case is a stream with just a single "line" `x ... (Return r)`. Because its `st` is `Return r` (assuming no monadic effects), neither of the special cases match and it becomes `x >> Chunk "\n" (Empty r)` by falling through to the third default case.  The result can only match the second of the special case patterns, provided `x` was actually `Empty (Return r)`.  It cannot match the first special case, there is no way to produce an empty initial chunk via either `cons'` or the second pattern, or from some non-empty `x`, so the first pattern is dead code.
- The next is a two line stream `Step (y ... (Step (x ... (Return r)))`, now the extant `unlines` will potentially match the second special pattern, if `x` was `Empty (Return)` (sans monadic effects).  And in this case we get `y >> Chunk "\n" (Empty r)` without a newline appended to `y`, and perhaps again end up in the second case when `y` is just `Empty`.

#### TLDR

Bottom line, the unpatched code turns 3 explicitly `Empty` lines: `Step $ Empty $ Step $ Empty $ Step $ Empty $ Return r` into a single line instead of the expected three lines:
```
-- Setup, prepare two streams of 3 empty lines:
-- The lines of `str1` are just `Empty (Step k)`, with `k` the next line
-- The lines of `str2` are `Chunk "" (Empty (Step k))`, with `k` the next line
--
λ> import qualified Streaming.Internal as S
λ> import qualified Streaming.ByteString.Internal as Q
λ> import qualified Streaming.ByteString.Char8 as Q
λ> let eof = S.Return ()
λ> let empty1 k = S.Step $ pure k
λ> let empty2 k = S.Step $ Q.Chunk mempty $ pure k
λ> let str1 = head $ drop 3 $ iterate empty1 eof
λ> let str2 = head $ drop 3 $ iterate empty2 eof

-- Unpatched:
λ> Q.unlines str1
Chunk "\n" (Empty (()))
λ> Q.unlines str2
Chunk "" (Chunk "\n" (Chunk "\n" (Chunk "\n" (Empty (())))))

-- Patched:
λ> Q.unlines str1
Chunk "\n\n\n" (Empty (()))
λ> Q.unlines str2
Chunk "" (Chunk "\n" (Chunk "\n" (Chunk "\n" (Empty (())))))

-- Potential revision (likely not warranted):
λ> Q.unlines str1
Chunk "\n\n\n" (Empty (()))
λ> Q.unlines str2
Chunk "\n\n\n" (Empty (()))
```

Potential revision avoids the nested `Chunk "\n"` constructs when the input stream contains empty chunks as `Chunk "" $ Empty r` rather than just `Empty r`, but I am not sure this is worthwhile. The library generally tries to avoid creating empty chunks.  So  code to optimise this case (below) is probably unwarranted:
```haskell
unlines :: Monad m => Stream (ByteStream m) m r ->  ByteStream m r
unlines = loop where
  loop str =  case str of
    Step bstr   -> case bstr of
        Chunk c  cs | B.null c -> unlines $ Step cs
                    | otherwise -> bstr >>= cons' '\n' . unlines
        Go m       -> Go $ (unlines . Step) <$> m
        Empty r    -> cons' '\n' $ unlines r
    Effect m  -> Go (fmap unlines m)
    Return r -> Empty r
```

File: lib/Streaming/ByteString/Internal.hs (hunk 4 above)
Comment (vdukhovni) [reply lib/Streaming/ByteString/Internal.hs:320]: Perhaps the issue is a concern around the efficiency of `toList` from `Streaming.Prelude`:
```
toList :: Monad m => Stream (Of a) m r -> m (Of [a] r)
toList = fold (\diff a ls -> diff (a: ls)) id (\diff -> diff [])
```
This builds a chain of "prepend" thunks, that are then applied to `[]`.  Is this a costly construction to run on a list of 4k elements? I'm not changing this, we can look into whether a larger number is warranted later...

File: lib/Streaming/ByteString.hs (hunk 1 above)
Comment (fosskers) [reply lib/Streaming/ByteString.hs]: Yes please :)

File: lib/Streaming/ByteString.hs (hunk 2 above)
Comment (fosskers) [reply lib/Streaming/ByteString.hs:437]: Considering how proactive you've been with benchmarking, it might make sense to make an official `bench` component for this project. Thoughts?

File: lib/Streaming/ByteString/Char8.hs (hunk 3 above)
Comment (fosskers) [reply lib/Streaming/ByteString/Char8.hs:531]: Thanks for that detailed explanation.

**Comment by fosskers** [comment]
We can consider the benchmarking issue in a later PR. For now, let's kill those "signature comments" and then merge this.

Oh, would you also mind updating the CHANGELOG?

File: lib/Streaming/ByteString.hs (hunk 2 above)
Comment (vdukhovni) [reply lib/Streaming/ByteString.hs:437]: > Considering how proactive you've been with benchmarking, it might make sense to make an official `bench` component for this project. Thoughts?

That would be cool, but I probably won't have the cycles to do it justice...

**Comment by vdukhovni** [comment]
> We can consider the benchmarking issue in a later PR. For now, let's kill those "signature comments" and then merge this.

Pushed a commit that drop the module export list signature comments.  I would however like to draw your attention to a couple of places where we still need to make decisions:

The `fold` and `fold_` signatures are `unnatural` in the "Word8" module, as a breaking change, we should probably switch them around.  Also the `uncons` issues could be folded into this PR, or done separately...

> 
> Oh, would you also mind updating the CHANGELOG?

We can do that once the full scope is known. Still a few minor issues to decide...

**Comment by vdukhovni** [comment]
Once this is almost ready, I'd like to squash the commits.

**Comment by fosskers** [comment]
How are the `fold` signatures unnatural?

**Comment by vdukhovni** [comment]
> How are the `fold` signatures unnatural?

Well, this was the one thing that side-by-side signatures in the export list were good for. :-)
For almost all the functions `somefunc` is the one the retains the `r` value in the result structure and `somefunc_` is the one that drops it. However, with `fold` in the non-`Char8` module, it is the other way around:
```
$ git grep '^fold_* *:: '  | sed 's/:: Monad m => /:: /'
ByteString.hs:fold :: (x -> Word8 -> x) -> x -> (x -> b) -> ByteStream m () -> m b
ByteString.hs:fold_ :: (x -> Word8 -> x) -> x -> (x -> b) -> ByteStream m r -> m (Of b r)
ByteString/Char8.hs:fold_ :: (x -> Char -> x) -> x -> (x -> b) -> ByteStream m () -> m b
ByteString/Char8.hs:fold :: (x -> Char -> x) -> x -> (x -> b) -> ByteStream m r -> m (Of b r)
```

I think that while we're doing breaking change housekeeping, the first two should be switched.

**Comment by fosskers** [comment]
I agree, let's move ahead with that.

**Comment by vdukhovni** [comment]
> I agree, let's move ahead with that.

I added this as a *fixup* commit and also including the `toStrict_` signature change.  Next on the agenda is `uncons`. Any chance I can bundle that into #42 also?

**Comment by vdukhovni** [comment]
Squashed and added CHANGELOG entries, almost ready to merge, modulo including the `uncons`-related fixes in this PR.

**Comment by fosskers** [comment]
Feel free to go ahead with all the `uncons` fixes.

**Comment by vdukhovni** [comment]
I made the `uncons` changes as a separate unsquashed commit, should be easier to review.  I think that's everything now...

**Comment by fosskers** [comment]
Sweet, if all goes well I should be able to merge this tomorrow morning (it's 4pm here).

**Comment by fosskers** [comment]
Thank you very much for this!
